Sealed trait in Array can pickle but cannot unpickle. Tag X not recognized, looking for one of: Y, X

eed3si9n opened the pull request:
This is a bug report with a failing test case. Don't merge.
## steps

```
scala-pickling> testOnly scala.pickling.test.sealedtraitinarray*
```
## problem

```
[info] SealedTraitInArrayTest:
[info] - sealed trait in Array *** FAILED ***
[info]   scala.pickling.PicklingException: Tag scala.pickling.test.sealedtraitinarray.Fruit not recognized, looking for one of: scala.pickling.test.sealedtraitinarray.Apple, scala.pickling.test.sealedtraitinarray.Orange
[info]   at scala.pickling.test.sealedtraitinarray.SealedTraitInArrayTest$$anonfun$1$ScalaPicklingTestSealedtraitinarrayFruitUnpickler$macro$24$2$.unpickle(sealed-trait-in-array.scala:20)
[info]   at scala.pickling.pickler.TravPickler$$anon$1.unpickle(Iterable.scala:65)
[info]   at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
[info]   at scala.pickling.pickler.TravPickler$$anon$1.unpickleEntry(Iterable.scala:18)
[info]   at scala.pickling.functions$.unpickle(functions.scala:11)
[info]   at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
[info]   at scala.pickling.test.sealedtraitinarray.SealedTraitInArrayTest$$anonfun$1.apply$mcV$sp(sealed-trait-in-array.scala:20)
[info]   at scala.pickling.test.sealedtraitinarray.SealedTraitInArrayTest$$anonfun$1.apply(sealed-trait-in-array.scala:15)
[info]   at scala.pickling.test.sealedtraitinarray.SealedTraitInArrayTest$$anonfun$1.apply(sealed-trait-in-array.scala:15)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
```
## notes

```
JSONPickle({
  "$type": "scala.Array[scala.pickling.test.sealedtraitinarray.Fruit]",
  "elems": [
    {
    "$type": "scala.pickling.test.sealedtraitinarray.Apple",
    "kind": "x"
  },
    {
    "$type": "scala.pickling.test.sealedtraitinarray.Orange",
    "kind": "x"
  }
  ]
})
```

eed3si9n committed:
Sealed trait in Array can pickle but cannot unpickle

[info]   scala.pickling.PicklingException: Tag
scala.pickling.test.sealedtraitinarray.Fruit not recognized, looking
for one of: scala.pickling.test.sealedtraitinarray.Apple,
scala.pickling.test.sealedtraitinarray.Orange

jsuereth commented:
You should check to make sure the hintTag isn't getting pinned in this case: https://github.com/scala/pickling/blob/0.10.x/core/src/main/scala/pickling/pickler/Iterable.scala#L30-L35

jsuereth commented:
Actually, that shouldn't matter...

jsuereth commented:
Here's the issue:   We ignore the keyTag returend from beginEntry => https://github.com/scala/pickling/blob/0.10.x/core/src/main/scala/pickling/pickler/Iterable.scala#L64-L65

I think I have a fix for this.   It's a bit hairy to understand the details though, as it rubs the "hintStaticallyElided" rough-edges in the picklers, where those hints are 100% about efficiency or human-readability.  We should aim for 100% correctness and backport human readability on that.

jsuereth committed:
Fix iterable pickler to correctly pass hint tags read at runtime.

There was no reason for iterable pickler to devaite from unpickleEntry
default, so we remove the custom entry tag handling, fixing the issue.

jsuereth commented:
@eed3si9n I submitted a fix for this issue.   Please review.

eed3si9n commented:
@jsuereth Your part LGTM. Removing (wip) from the title.
